scriptoocloud/src/test/java/edu/asu/diging/scriptoocloud/core/service/impl/FileSystemServiceTest.java

```diff
@@ -1,72 +1,128 @@
 import org.mockito.MockitoAnnotations;
+import org.springframework.test.context.ContextConfiguration;
+import org.springframework.test.context.junit.jupiter.SpringExtension;
+import org.springframework.test.context.support.AnnotationConfigContextLoader;
 
-/*
- * 
- * 
+/**
  * @author Jason Ormsby
-*/
-
+ * @author John Coronite
+ */
+@ExtendWith(SpringExtension.class)
+@ContextConfiguration(classes = AnnotationConfigContextLoader.class)
 class FileSystemServiceTest {
 
-    /*
-     * New temporary directory generated for every test
-     */
-    @TempDir
-    protected Path folder;
+    private static final String ID = "1";
 
     @InjectMocks
-    private FileSystemService serviceToTest;
-    
+    private FileSystemService fileSystemService;
+
+    @TempDir
+    protected Path path;
+
+    private String username;
+
+    private String type;
+
     @BeforeEach
     public void init() {
-        MockitoAnnotations.initMocks(this);   
+        MockitoAnnotations.initMocks(this);
+        initPathSegments();
+    }
+
+    private void initPathSegments() {
+        String pathString = path.toString();
+        String[] segments = pathString.split(File.separator);
+        StringBuilder stringBuilder = new StringBuilder();
+        for (int i = 0; i < segments.length - 1; i++) {
+            stringBuilder.append(segments[i]);
+            if (i != segments.length - 2)
+                stringBuilder.append(File.separator);
+        }
+        username = stringBuilder.toString();
+        type = segments[segments.length - 1];
     }
-    
+
     @Test
-    public void test_deleteDirectoryOrFile_deleteFile() throws IOException {
-        File file = new File(folder+"myfile");
-        file.createNewFile();
-        
-        Assertions.assertEquals(true, file.exists());
-        
-        serviceToTest.deleteDirectoryOrFile(file);
-        
-        Assertions.assertEquals(false, file.exists());
-    }   
-    
+    public void test_addDirectories() throws FileSystemStorageException, IOException {
+        fileSystemService.addDirectories(username, type, ID);
+        File file = Paths.get(path + ID).toFile();
+        Assertions.assertTrue(file.createNewFile());
+        fileSystemService.deleteDirectoryOrFile(file);
+        Assertions.assertFalse(file.exists());
+    }
+
     @Test
-    public void test_deleteDirectoryOrFile_deleteFolderAndContents() throws IOException {
-        Files.write(folder.resolve("myfile"), "abc".getBytes());
-        File file = folder.resolve("myfile").toFile();
-        
-        Assertions.assertEquals(true, file.exists());
-        Assertions.assertEquals(true, folder.toFile().exists());
-        
-        serviceToTest.deleteDirectoryOrFile(folder.toFile());
-        
-        Assertions.assertEquals(false, file.exists());
-        Assertions.assertEquals(false, folder.toFile().exists());
-    }   
-    
+    public void test_addDirectories_failed() {
+        Assertions.assertThrows(NullPointerException.class,
```

Comment:
The `NullPointerExceptions` for both test cases are thrown in the `FileSystemService` `createPath()` method - in particular, Java's `Paths.get(null)` therein throws a `NullPointerException`.  I couldn't quite figure out how to test failure for `addDirectories()` and similarly, failure for `createPath()`.  Every combination of strings I tried is cleaned in the `createPath()` method and returns successfully.  I remember you mentioned a while back that it is generally bad practice to catch `NullPointerException`s, so I'm wondering if and how `createPath()` should be altered to deal with null inputs, and/or how to test the method(s) appropriately?

Reply:
I think I would have addDirectories return a boolean (true if directory was created, otherwise false). Then check in the method if the passed values are all non null. If username or type are null, return false, which is the condition your test can check (plus if the folder was created maybe). similar for the `createPath` method. I would actually probably allow to just pass in `username` and the other values null to just create a user folder. And then check if at least username is non null. If it is null, return null which your test can check for. Does this make sense?